nupic/research/frameworks/vernon/core_mixins/step_based_logging.py

```diff
@@ -0,0 +1,153 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2020, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import math
+from collections import defaultdict
+
+__all__ = [
+    "StepBasedLogging",
+]
+
+
+class StepBasedLogging:
```

Comment:
This looks great. In the future I was wondering if we could refactor in such a way that user can define the level he wants to log, eg: step, epoch, task, bundle of tasks (as in meta learning). I haven't given a lot of thought about this, and I don't know a good approach to do this given our current framework. But maybe we can pingpong ideas some day.